toy_problem/src/TaskAllocation.cpp

```diff
@@ -0,0 +1,672 @@
+
+    // Compute time taken to reach charging waypoint from current location
+    state->waypoint = initial_state.charging_waypoint;
+    const auto start_time = std::chrono::steady_clock::now() +
+      rmf_traffic::time::from_seconds(initial_state.finish_time);
+    rmf_traffic::agv::Planner::Start start{
+      start_time,
+      initial_state.waypoint,
+      0.0};
+
+    rmf_traffic::agv::Planner::Goal goal{initial_state.charging_waypoint};
+
+    const auto result = _planner->setup(start, goal);
+    state->finish_time += result.initial_cost_estimate();
+
+    // TODO Check if robot has charge to make it back to its charging dock
+    //
+
+    double battery_soc = initial_state.battery_soc;
+    double delta_soc = _charge_soc - battery_soc;
+    assert(delta_soc >= 0.0);
+    double time_to_charge =
+      (3600 * delta_soc * _battery_system->nominal_capacity()) / _battery_system->charging_current();
+
+    state->finish_time += time_to_charge;
+
+    return state;
+  }
+
+private:
+  // fixed id for now
+  std::size_t _id = 42;
+  std::shared_ptr<rmf_battery::agv::BatterySystem> _battery_system;
+  std::shared_ptr<rmf_traffic::agv::Planner> _planner;
+  // soc to always charge the battery up to
+  double _charge_soc = 1.0;
+ 
+};
+
+// ============================================================================
+class DeliveryTaskRequest : public TaskRequest
+{
+public:
+  using Planner = rmf_traffic::agv::Planner;
+  DeliveryTaskRequest(
+    std::size_t id,
+    std::size_t pickup,
+    std::size_t dropoff,
+    std::shared_ptr<Planner> planner)
+  : _id(id),
+    _pickup_waypoint(pickup),
+    _dropoff_waypoint(dropoff),
+    _planner(planner)
+  {
+    // Do nothing
+  }
+
+static ConstTaskRequestPtr make(
+  std::size_t id,
+  std::size_t pickup,
+  std::size_t dropoff,
+  std::shared_ptr<Planner> planner)
+{
+  return std::make_shared<DeliveryTaskRequest>(id, pickup, dropoff, planner);
+}
+
+std::size_t id() const final
+{
+  return _id;
+}
+
+rmf_utils::optional<RobotState> estimate(const RobotState& initial_state) const final
+{
+  rmf_utils::optional<RobotState> state = initial_state;
+  state->waypoint = _dropoff_waypoint;
+  const auto now = std::chrono::steady_clock::now();
+
+  const auto start_time = now +
+    rmf_traffic::time::from_seconds(initial_state.finish_time);
+  Planner::Start start{
+    start_time,
+    initial_state.waypoint,
+    0.0};
+
+  Planner::Goal goal{_pickup_waypoint};
+  
+  // First move to waypoint on graph
+  const auto result_to_pickup = _planner->setup(start, goal);
+  const double cost_estimate = result_to_pickup.initial_cost_estimate();
+  state->finish_time += cost_estimate;
+
+  Planner::Start start_2{
+    rmf_traffic::time::apply_offset(start_time, cost_estimate),
+    _pickup_waypoint,
+    0.0};
+
+  Planner::Goal goal_2{_dropoff_waypoint};
+
+  const auto result_to_dropoff = _planner->setup(start_2, goal_2);
+  state->finish_time += result_to_dropoff.initial_cost_estimate();    
```

Comment:
I was thinking about this portion. this indicates the accumulation of "cost". As far as I know, having `initial_state.finish_time` has already done the job of accumulating the cost for each task which is in sequence. In first glance, this might create a double count issue. Similar to above, this might create an issue. :thinking: